### ISSUE TITLE
Self-signing HTTPS certificate guide/documentation

### ISSUE THREAD

**Issue by EntropySmoke** [AdguardTeam/AdGuardHome#2266]
### Prerequisites
- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Problem Description
AdGuard Home provides links to letsencrypt.org as a way of obtaining free HTTPS certificates, but letsencrypt.org does not provide such certificates for local domains. For local AdGuard Home DNS servers, it makes more sense to create self-signed HTTPS certificates. Such certificates are considered "invalid" due to being self-signed, but they still provide an extra layer of security for local AdGuard Home DNS servers. Some cable internet ISP's use such self-signed HTTPS certificates for their cable modems. AdGuard Home accepts such "invalid" certificates. 

### Proposed Solution
Provide documentation and/or links to documentation on how to generate self-signed HTTPS certificates.

### Alternatives Considered
Provide OpenSSL tools that can generate self-signed HTTPS certificates.

**Comment by Aikatsui** [comment]
Bot: Wontfix
Aikatsui: Ta-da!..

**Comment by SignFinder** [comment]
at minimum the information how to correct add files of letsencrypt certificate to ADGuard Home are necessary.

**Comment by subnut** [comment]
This issue hasn't been resolved... why close it? 😕

**Comment by X448NAX** [comment]
Caddy is a very user friendly way to automate this, if you are interested I can write a tutorial. But essentially if you install Caddy on the same machine as AdGuard Home and tell it to set a reverse proxy to the correct port on localhost for AdGuard Home, it'll automatically set up a CA and generate self-signed certificates. You'll then just need to trust the CA manually on your other devices and that's it.

Since Caddy will be serving the admin interface as a reverse proxy you don't even need to set up the certificates inside AdGuard Home this way unless you want local DNS traffic to use encrypted DNS. You can do this if you want to regardless. But even without further config of AdGuard Home itself, Caddy will encrypt the local traffic between your machine and the admin console via HTTPS which I assume is what most people want to protect.

If you do want the self-signed certs to provide DoH and DoT you can put the path for those certs into AdGuard Home, just make sure the user AdGuard Home runs under has permissions to access the certs.

**Comment by jonas-w** [comment]
# Got it working without a reverse proxy
### How to Setup Encryption without reverse proxy
- I followed [this guide on how to create own Certificate Authority](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/)
- Then installed the certificates on my raspberry pi where AdGuard Home is installed. Instructions are [here](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/#adding-root-cert-linux-keychain)
- After running `sudo update-ca-certificates` restart your AdGuardHome service (`systemctl restart AdGuardHome`)
- Now copy `your-domainname.crt` and `yourdomainname.key` to e.g. `/opt/AdGuardHome/`

### Installing the certificates on AdGuardHome
1. Login on the Web Panel
2. Go to Settings → Encryption Settings
3. Enable Encryption
4. Change `Server Name` to your-domainname
5. Change other settings however you want / you probably want to enable https redirect to always use https
6. Now set the "Certificate Path" to e.g. `/opt/AdGuardHome/your-domainname.crt`
7. And change the "Private key path" to e.g. `/opt/AdGuardHome/your-domainname.key`
8. Last but not least click **Save Configuration**

As you can see your certificate is **NOT invalid** and will be trusted by AdGuardHome!

Now you can access your adguard home via https with https://your-domainname:YOUR-HTTPS-PORT without a reverse proxy

**Comment by gl4d** [comment]
> # Got it working without a reverse proxy
> ### How to Setup Encryption without reverse proxy
> * I followed [this guide on how to create own Certificate Authority](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/)
> * Then installed the certificates on my raspberry pi where AdGuard Home is installed. Instructions are [here](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/#adding-root-cert-linux-keychain)
> * After running `sudo update-ca-certificates` restart your AdGuardHome service (`systemctl restart AdGuardHome`)
> * Now copy `your-domainname.crt` and `yourdomainname.key` to e.g. `/opt/AdGuardHome/`
> 
> ### Installing the certificates on AdGuardHome
> 1. Login on the Web Panel
> 2. Go to Settings → Encryption Settings
> 3. Enable Encryption
> 4. Change `Server Name` to your-domainname
> 5. Change other settings however you want / you probably want to enable https redirect to always use https
> 6. Now set the "Certificate Path" to e.g. `/opt/AdGuardHome/your-domainname.crt`
> 7. And change the "Private key path" to e.g. `/opt/AdGuardHome/your-domainname.key`
> 8. Last but not least click **Save Configuration**
> 
> As you can see your certificate is **NOT invalid** and will be trusted by AdGuardHome!
> 
> Now you can access your adguard home via https with https://your-domainname:YOUR-HTTPS-PORT without a reverse proxy

I've followed the instructions, AdGuard marks the certificate & private key as "valid" but chrome is still displaying NET::ERR_CERT_COMMON_NAME_INVALID / shows the certificate as not valid & connection not secure.

I found the solution, in case anyone gets the same error in the future: 
it was because i wasn't using a domain/dns but a local IP, [here](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/#creating-ca-signed-certificates) the tutorial tells you to create a *.ext file with the following content:

> authorityKeyIdentifier=keyid,issuer
> basicConstraints=CA:FALSE
> keyUsage = digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment
> subjectAltName = @alt_names
> 
> [alt_names]
> DNS.1 = hellfish.test

instead of 
> subjectAltName = @alt_names
> 
> [alt_names]
> DNS.1 = hellfish.test

you just need

>subjectAltName = IP:192.168.1.X

final *.ext file: 

> authorityKeyIdentifier=keyid,issuer
> basicConstraints=CA:FALSE
> keyUsage = digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment
> subjectAltName = IP:192.168.1.X

**Comment by MNLierman** [comment]
F—king ridiculous bullshit that community developers in charge of the things we know and love have shit for options and miss the mark on basic things we've request. Harsh but you know what, pretty sick of this trend. As someone who generally contributed to every free thing I use, it's disheartening to feel this disappointed. Like I said. Huge growing trend: Give users less, decide for them what they can and can't have, don't fix major bugs. 

Oh well for now. I'll be trying this solution posted above. I shouldn't have to do that tho, that's my point. Thanks to @jonas-w for posting this information, great to see good follow up like this indexed on Google.

**Comment by donluca** [comment]
> # Got it working without a reverse proxy
> ### How to Setup Encryption without reverse proxy
> * I followed [this guide on how to create own Certificate Authority](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/)
> * Then installed the certificates on my raspberry pi where AdGuard Home is installed. Instructions are [here](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/#adding-root-cert-linux-keychain)
> * After running `sudo update-ca-certificates` restart your AdGuardHome service (`systemctl restart AdGuardHome`)
> * Now copy `your-domainname.crt` and `yourdomainname.key` to e.g. `/opt/AdGuardHome/`
> 
> ### Installing the certificates on AdGuardHome
> 1. Login on the Web Panel
> 2. Go to Settings → Encryption Settings
> 3. Enable Encryption
> 4. Change `Server Name` to your-domainname
> 5. Change other settings however you want / you probably want to enable https redirect to always use https
> 6. Now set the "Certificate Path" to e.g. `/opt/AdGuardHome/your-domainname.crt`
> 7. And change the "Private key path" to e.g. `/opt/AdGuardHome/your-domainname.key`
> 8. Last but not least click **Save Configuration**
> 
> As you can see your certificate is **NOT invalid** and will be trusted by AdGuardHome!
> 
> Now you can access your adguard home via https with https://your-domainname:YOUR-HTTPS-PORT without a reverse proxy

Unfortunately this is not working for me as I get an error saying

"validating certificate pair: certificate does not verify: x509: certificate is not valid for any names, but wanted to match aurora"

on the adguardhome web UI

**Comment by jorti** [comment]
> Unfortunately this is not working for me as I get an error saying
> 
> "validating certificate pair: certificate does not verify: x509: certificate is not valid for any names, but wanted to match aurora"
> 
> on the adguardhome web UI

Sounds to me like the configured server name doesn't match the SAN of the certificate.

**Comment by donluca** [comment]
I've managed to sort it out and fix the issues, I don't know how those steps could have possibly worked in the past.
Anyway...

```
openssl genrsa -out ca-key.pem 2048
openssl req -new -x509 -days 365 -key ca-key.pem -out ca.pem -subj "/CN=localhost" -addext "subjectAltName = DNS:localhost"
cp ca.pem /usr/local/share/ca-certificates/ca.crt
update-ca-certificates
service adguardhome restart
cp /usr/local/share/ca-certificates/ca.crt /var/lib/adguardhome/localhost.crt
cp ca-key.pem /var/lib/adguardhome/localhost.pem
cd /var/lib/adguardhome/
chown adguardhome:adguardhome localhost.*
```

Change `/var/lib/adguardhome` to whatever path you have AdGuard Home installed.

Then go to the webGUI of Adguard, Settings, Encryption, change Server Name to localhost, set the paths accordingly and you're off to the races.

One important thing: here I've set up the key without a passphrase otherwise I get an error from Adguard.
This is generally discouraged so it's up to you what to do with it.

Hope this helps.

**Comment by K-MeLeOn** [comment]
@donluca Key file unencrypted is a common practice if the app dosn't use `PKCS#12`. In a properly configured Self-Signed CA you must provide a `CLR` before generating a certificate (to include it in the final certificate). This allow you to revoke any certificate. In terms of risk, there is only one if your AGH certifcate is a wildcart one, otherwise you provide only `IP` and `CN` according to your ADG instance. If you have multiple docker above the same `IP`, user an `internal reverse-proxy` like Traefik with `pathrought TLS` to resolve your apps with the DNS name over IP.

For those searching how to archieve a Self-Signed CA approval for AGH, as mentionned before, AGH use the trusted certificates store of the machine, if you havn't already added your `intermediate-ca`, AGH can't resolve the chain validation.

For docker compose file, you can do this to automate the process over redploying container (I'll indicate only usefull thing, update your docker-compose according to your environment) :

```yaml
services:
  agh:
    entrypoint:
      - /bin/sh
      - -c
      - |
        update-ca-certificates
        exec /opt/adguardhome/AdGuardHome --no-check-update -c /opt/adguardhome/conf/AdGuardHome.yaml -w /opt/adguardhome/work
    volumes:
      - ${AGH_PATH}/work/certs:/opt/adguardhome/work/certs:ro
      - ${AGH_PATH}/work/certs/intermediate-ca.pem:/usr/local/share/ca-certificates/intermediate-ca.pem:ro

```

You need to only provide the `intermediate-ca`, if you provide the `bundle-ca` with your intermediate(s) & root, it won't work (try it manually into the docker and see the `update-ca-certificates` result).

With that when you redeploy your AGH, it will automatically map your `intermediate-ca` and permform a `update-ca-certificates` BEFORE starting AGH to be available.

Please, if in the future the [entrypoint command](https://github.com/AdguardTeam/AdGuardHome/blob/4ca8533efc8a5bd60d4781b6afa696cfde0c356f/docker/Dockerfile#L62) change, don't forget to update it.

**Comment by gpz1100** [comment]
I've been using LE certs for local domains for some time now.  The internal domain is defined as i.domain.com.  LE cert is obtained for *.i.domain.com and i.domain.com using dns-01 challenge using your favorite LE client.

As this is a wildcard cert, it's valid for {any hostname}.i.domain.com. No need to add self signed certs to browsers.  Local servers with static ip's are defined in the custom filters section of adguard home so SNI works properly, except the target ip is rfc1918.
